sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/helpers/KinesisClientProxyStub.java

```diff
@@ -0,0 +1,194 @@
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.apache.beam.sdk.io.aws2.kinesis.enhancedfanout.AsyncClientProxy;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.services.kinesis.model.ListShardsRequest;
+import software.amazon.awssdk.services.kinesis.model.ListShardsResponse;
+import software.amazon.awssdk.services.kinesis.model.SequenceNumberRange;
+import software.amazon.awssdk.services.kinesis.model.Shard;
+import software.amazon.awssdk.services.kinesis.model.SubscribeToShardEvent;
+import software.amazon.awssdk.services.kinesis.model.SubscribeToShardEventStream;
+import software.amazon.awssdk.services.kinesis.model.SubscribeToShardRequest;
+import software.amazon.awssdk.services.kinesis.model.SubscribeToShardResponse;
+import software.amazon.awssdk.services.kinesis.model.SubscribeToShardResponseHandler;
+
+public class KinesisClientProxyStub implements AsyncClientProxy {
+  private final KinesisClientStubConfig config;
+  private final BlockingQueue<SubscribeToShardRequest> subscribeRequestsCollector;
+  private final AtomicInteger seqNumber;
+  private final AtomicInteger subscriptionsCountdown;
+  private final Function<KinesisClientStubShardState, Void> eventsSubmitter;
+
+  KinesisClientProxyStub(
+      KinesisClientStubConfig config,
+      AtomicInteger subscriptionsCountdown,
+      AtomicInteger seqNumber,
+      BlockingQueue<SubscribeToShardRequest> subscribeRequestsCollector,
+      Function<KinesisClientStubShardState, Void> eventsSubmitter) {
+    this.config = config;
+    this.seqNumber = seqNumber;
+    this.subscriptionsCountdown = subscriptionsCountdown;
+    this.subscribeRequestsCollector = subscribeRequestsCollector;
+    this.eventsSubmitter = eventsSubmitter;
+  }
+
+  @Override
+  public CompletableFuture<ListShardsResponse> listShards(ListShardsRequest listShardsRequest) {
+    if (listShardsRequest.shardFilter().shardId() == null) {
+      return CompletableFuture.completedFuture(
+          ListShardsResponse.builder()
+              .shards(buildShards(0, config.getInitialShardsCnt()))
+              .build());
+    } else {
+      return CompletableFuture.completedFuture(
+          ListShardsResponse.builder()
+              .shards(buildShards(config.getInitialShardsCnt(), config.getFinalShardId() + 1))
+              .build());
+    }
+  }
+
+  @Override
+  public CompletableFuture<Void> subscribeToShard(
+      SubscribeToShardRequest request, SubscribeToShardResponseHandler responseHandler) {
+    subscribeRequestsCollector.add(request);
+
+    return CompletableFuture.supplyAsync(
```

Comment:
@psolomin Regarding tests, there's definitely a problem with the stubbed behavior here. The future returned by supplyAsync completes instantly, there's nothing that would block / delay that. And that triggers an immediate resubscribe.

Give it a try like this:

```Java
package org.apache.beam.sdk.io.aws2.kinesis.enhancedfanout;

import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;
import software.amazon.awssdk.core.async.SdkPublisher;
import software.amazon.awssdk.services.kinesis.KinesisAsyncClient;
import software.amazon.awssdk.services.kinesis.model.SubscribeToShardEventStream;
import software.amazon.awssdk.services.kinesis.model.SubscribeToShardRequest;
import software.amazon.awssdk.services.kinesis.model.SubscribeToShardResponseHandler;

import javax.annotation.Nullable;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.atomic.AtomicInteger;

import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkNotNull;

// Just implements subscribeToShard, everything else will throw
public class StubbedKinesisAsyncClient implements KinesisAsyncClient {

  private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
  private final int publisherRateMs;
  private final Map<String, Deque<StubbedSdkPublisher>> stubbedPublishers = new HashMap<>();

  public StubbedKinesisAsyncClient(int publisherRateMs) {
    this.publisherRateMs = publisherRateMs;
  }

  /**
   * Stubs a subscribeToShard call with the provided events, optionally terminating with an error or
   * otherwise normally as soon as all events are delivered.
   */
  public CanFail stubSubscribeToShard(String shardId, SubscribeToShardEventStream... events) {
    StubbedSdkPublisher publisher = new StubbedSdkPublisher(events);
    stubbedPublishers.computeIfAbsent(shardId, id -> new ArrayDeque<>()).add(publisher);
    return publisher;
  }

  @Override
  public CompletableFuture<Void> subscribeToShard(
      SubscribeToShardRequest req, SubscribeToShardResponseHandler resp) {
    Deque<StubbedSdkPublisher> publishers =
        checkNotNull(stubbedPublishers.get(req.shardId()), "Not stubbed");
    StubbedSdkPublisher publisher = checkNotNull(publishers.poll(), "No more stubs");
    resp.onEventStream(publisher);
    return publisher.result;
  }

  @Override
  public void close() {
    scheduler.shutdown();
  }

  @Override
  public String serviceName() {
    return "kinesis";
  }

  public interface CanFail {
    void failWith(Throwable error);
  }

  private class StubbedSdkPublisher implements SdkPublisher<SubscribeToShardEventStream>, CanFail {
    final CompletableFuture<Void> result = new CompletableFuture<>();
    final SubscribeToShardEventStream[] events;
    @Nullable Throwable error = null;

    StubbedSdkPublisher(SubscribeToShardEventStream[] events) {
      this.events = events;
    }

    public void failWith(Throwable error) {
      this.error = error;
    }

    @Override
    public void subscribe(Subscriber<? super SubscribeToShardEventStream> subscriber) {
      AtomicInteger requested = new AtomicInteger();
      subscriber.onSubscribe(
          new Subscription() {
            {
              scheduler.schedule(this::publish, publisherRateMs, MILLISECONDS);
            }

            @Override
            public void request(long n) {
              requested.incrementAndGet();
            }

            @Override
            public void cancel() {
              result.complete(null);
            }

            int idx = 0;

            void publish() {
              if (!result.isDone() && idx < events.length) {
                if (requested.getAndUpdate(i -> Math.max(0, i - 1)) > 0) {
                  subscriber.onNext(events[idx++]);
                }
                scheduler.schedule(this::publish, publisherRateMs, MILLISECONDS);
              } else if (error != null) {
                subscriber.onError(error);
                result.completeExceptionally(error);
              } else {
                subscriber.onComplete();
                result.complete(null);
              }
            }
          });
    }
  }
}

```